### PR TITLE
Add zlib format support

### PR DESCRIPTION
**Pull request by QKaiser**
See #422

### CODE REVIEW

**Comment by QKaiser** [comment]
False positive zlib content is identified in squashfs integration tests samples. I need to find a way for false positive reduction.

**Comment by vlaci** [comment]
Yeah, these 2 byte patterns are well, not very specific :)

**Comment by vlaci** [comment]
> > Yeah, these 2 byte patterns are well, not very specific :)
> 
> It's weird because it's a true positive (metadata in squashfs being compressed with zlib), but I don't understand why the squashfsv3 handler does not have priority over zlib in that the zlib chunk is within a larger squashfs chunk.

You can check what happens, when you disable other handlers. It is possible that the squashfs chunk is discarded for some reason, for example because the file is truncated.

File: unblob/handlers/compression/zlib.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        with File.from_path(inpath) as f:
+            while not decompressor.eof:
+                outpath.write_bytes(decompressor.decompress(f.read(DEFAULT_BUFSIZE)))
+
+
+class ZlibHandler(Handler):
+    NAME = "zlib"
+
+    PATTERNS = [
+        HexString("78 9c"),  # default compression
```
Comment (kissgyorgy) [inline unblob/handlers/compression/zlib.py:26]: What about 78 01?

Comment (QKaiser) [reply unblob/handlers/compression/zlib.py:27]: nice catch, fixed.

**Comment by vlaci** [comment]
I've been thinking that would it make sense for the time being to detect zlib only at the begining of files? Any legitimate file can contain a bunch zlib compressed or zlib like data streams which may or may not make sense to extract and detect. Even if we wouldn't have the current issue it colliding with squashfs.

**Comment by QKaiser** [comment]
This would not be sufficient for some samples that we have where the firmware is a zlib stream in the middle of another file.

Another way would be to check that the zlib chunk is bigger than some arbitrary value and emit an InvalidInput error otherwise.

**Comment by QKaiser** [comment]
The integration test failure was due to a bug in our squashfs ddwrt handler. I fixed it in https://github.com/onekey-sec/unblob/pull/426. As for zlib streams in multi-media files we can use the default skip list. 

I ran this branch against a bunch of firmwares, the only file type we should add to the skip list is .woff font files.

**Review by kissgyorgy — CHANGES_REQUESTED** [review]
Also please update the documentation too!

**Comment by QKaiser** [comment]
Fixed the docs, but still had a problem with DMG archives. Given that they can be made of multiple zlib streams, without a custom header but a custom footer. The fix I wrote checks if the file contains a DMG magic and returns early. This may be generalized in the future if similar cases arise.

File: docs/formats.md [hunk 2]
```diff
@@ -66,6 +66,7 @@ For compression formats, metadata cannot be preserved, as this information in mo
 | LZMA          | [compression/lzma.py][lzma-handler]         | [`7z`][lzma-extractor]          |
 | LZO           | [compression/lzo.py][lzo-handler]           | [`lzop`][lzo-extractor]         |
 | XZ            | [compression/xz.py][xz-handler]             | [`7z`][xz-extractor]            |
+| ZLIB          | [compression/zlib.py][zlib-handler]         | [`ZlibExtractor` custom code]zlib-extractor]        |
```
Comment (kissgyorgy) [inline docs/formats.md:69]: There is a missing `[` so it's not rendered correctly.  
You can spot these kinds of errors by using `mkdocs serve` and looking at the generated documentation.

File: unblob/handlers/compression/zlib.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        HexString("78 5e"),  # compressed
+    ]
+
+    EXTRACTOR = ZlibExtractor()
+
+    def calculate_chunk(self, file: File, start_offset: int) -> Optional[ValidChunk]:
+
+        eof = file.size()
+        for pattern in DMGHandler.PATTERNS:
+            if re.search(pattern.as_regex(), file[eof - 512 : eof]):  # noqa: E203
```
Comment (kissgyorgy) [inline unblob/handlers/compression/zlib.py:41]: You can just say "the last 512 bytes" without using the size of the file.
```suggestion
            if re.search(pattern.as_regex(), file[-512:]):
```

File: docs/formats.md [hunk 4]
```diff
@@ -86,6 +87,8 @@ For compression formats, metadata cannot be preserved, as this information in mo
 [lzo-extractor]: https://github.com/onekey-sec/unblob/blob/3008039881a0434deb75962e7999b7e35aca8271/unblob/handlers/compression/lzo.py#L78
 [xz-handler]: https://github.com/onekey-sec/unblob/blob/main/unblob/handlers/compression/xz.py
 [xz-extractor]: https://github.com/onekey-sec/unblob/blob/3008039881a0434deb75962e7999b7e35aca8271/unblob/handlers/compression/xz.py#L173
+[zlib-handler]: https://github.com/onekey-sec/unblob/blob/main/unblob/handlers/compression/zlib.py
```
[the remark this answers is not in the document]
Comment (QKaiser) [reply docs/formats.md]: Done.

File: unblob/handlers/compression/zlib.py (hunk 3 above)
Comment (QKaiser) [reply unblob/handlers/compression/zlib.py]: 👍

File: docs/formats.md (hunk 2 above)
Comment (QKaiser) [reply docs/formats.md]: good to know ! fixed.

**Review by kissgyorgy — APPROVED** [review]

**Pull request merged**